Fixes the missing exception throw action when an api request is executed, like BucketExistsAsync, etc.

ebozduman opened the pull request:
Fixes #1118
and probably #1114, #1121, #1131 and #1132, ....
There may be some more issues filed for this very same issue.
If I find more, I'll add them here.
#1120 (The PR that tried to fix the issue)

ramondeklein commented on Minio/RequestExtensions.cs:
```diff
@@ -145,9 +145,13 @@ await requestMessageBuilder.ResponseWriter(responseResult.ContentStream, cancell
+                throw;
+            }
+
+            responseResult = new ResponseResult(request, ex);
```
This code will never be executed. Because `responseResult` is set at line 93, the previous `if` statement will always be true and the exception will be rethrown.

ebozduman replied:
Done!

ebozduman replied:
Outdated!

ramondeklein commented on Minio/RequestExtensions.cs:
```diff
@@ -145,9 +145,13 @@ await requestMessageBuilder.ResponseWriter(responseResult.ContentStream, cancell
                 responseResult.Exception = ex;
-            else
-                responseResult = new ResponseResult(request, ex);
+                throw;
```
It makes no sense to set `responseResult.Exception` in the previous line and now rethrow the exception. The caller of this function will never receive the `responseResult`.

ebozduman replied:
removed!

ramondeklein commented on Minio/RequestExtensions.cs:
All code from lines 147-155 can be replaced by `throw;` and you would have had the same functionality as you implemented right now.

ebozduman replied:
Done!

ramondeklein reviewed (changes requested):
I can't really sign off on this. Code can be simplified, but even if you do, this core method changes semantics significantly and may have serious impact on API usage. Instead of returning an error response, this throws an exception instead.

There are way too much issues with this SDK. If we get reports that even simple `BucketExists` (by far the simplest operation in the SDK) isn't reliable. We should have a broader discussion what to do with this SDK.

rhegner commented:
@ebozduman even with your suggested changes, that entire error handling does not look very trustworthy to me.

@ramondeklein already mentioned some issues.

I didn't dig very deep in the code base, but it seems to me that the root cause of all the evil is that there is a mix of _throwing_ exceptions and _returning_ (unthrown) exception objects within the `ResponseResult`. For example, with your latest code change, in most cases exception are (re-)thrown. But if the `responseResult` variable is null in line 147, the exception will not be thrown, but instead returned. The fact that these two approaches are mixed, makes understanding (and maintaining) the entire error handling logic pretty difficult and fragile. [The fact that error handling is broken in various ways across multiple versions of the library is the best proof for that.]

My suggestion would be this (again, I don't really know the code base, so I might miss something):
Remove the `Exception` field in `ResponseResult` completely and transport exceptions only by throwing them. I think this would streamline the entire exception handling logic (and I think the required changes are limited to only a small part of the codebase).

---

Additional thoughts:

I could imagine that the original rationale for transporting exceptions inside of `ResponseResult` instead of throwing was to facilitate the idea of invoking "error handlers".
BUT: That mechanism seems to be broken anyways, see `HandleIfErrorResponse` method. If there is an exception, the method throws in line 386, so the handlers will never be invoked in line 391.

So, again, I think the entire error handling needs simplification and streamlining. Let all exceptions throw. If you want to keep the concept of error handlers, invoke them from within a catch block in an appropriate place, and then re-throw.

COCPORN commented:
Just merge the obvious fix.

This is embarrassing.

rhegner commented:
@ebozduman these are the results of my testbench:

```
// Minio Version                                6.0.0   6.0.1   6.0.2   6.0.3   PR
// ------------------------------------------------------------------------------------------------
await TestExistsWithExistingObject();       //  ok      ok      ok      ok      ok
await TestExistsWithNonexistingObject();    //  ok      NOK     ok      ok      ok
await TestExistsFromInvalidEndpoint();      //  ok      ok      NOK     NOK     different exception
await TestDownloadWithExistingObject();     //  ok      ok      ok      ok      ok
await TestDownloadWithNonexistingObject();  //  ok      ok      ok      ok      ok
await TestDownloadFromInvalidEndpoint();    //  ok      ok      NOK     NOK     different exception
```

Previous versions did throw a `ConnectionException`, with your latest code a `HttpRequestException` is thrown.
In our environment it would probably be ok to receive a different exception in these cases, but generally I would consider this a breaking change.

ebozduman commented:
@rhegner,
Thank you very much for testing the fix.
I'll look into the regressed exceptions.

harshavardhana commented:
> @rhegner, Thank you very much for testing the fix. I'll look into the regressed exceptions. I am also running our functional tests against 3 scenarios:
> 
> * Nonexisting endpoint
> * Invalid access  key
> * Invalid secret key
>   I'll work on all the anomalies and regressions I see in these tests.

@ebozduman is this done?

ebozduman commented:
@harshavardhana 

It is imcomplete!

Replaced the existing hack fix and cleaned up all the extra code.
Also made changes according to Ramon's suggestions.
But since this change touches everybody, testing took quite sometime.
There is another tune up that needs to be done for a test, but before finishing and testing the required change, I switched to the AI  storage testing. So, it is still incomplete.

jonahbohlmann commented:
@ebozduman
Maybe I'm misunderstanding this, but from my side this issue makes the library unstable and unusable.
We have already encountered unexpected scenarios in our applications several times because there are no clear error messages.

Is there anything the community can do to help resolve the issue in the short term? Maybe it is possible to release a beta version so we can test current fixes?

ebozduman commented:
@strelok372 
Thank you for your code review approval, but the fix requested was to fix the whole error handling code and logic and to clean  up the old code. So, this fix is not good enough. 
The desired fix is 80% done and verification and testing still in progress as the fix is quite big.
I've also talked and asked about marking the package as unstable until this issue is fixed, or some other solution is proposed.

strelok372 commented:
Okay, I understand. I think it's a good idea to mark it as unstable, because I spent some time to find an issue and should return to 6.0.1 after all. It would be useful to notice other people to save their time.00:16, 4 февраля 2025 г., ebozduman ***@***.***>:
@strelok372
Thank you for your code review approval, but the fix requested was to fix the whole error handling code and logic and to clean  up the old code. So, this fix is not good enough.
The desired fix is 80% done and verification and testing still in progress as the fix is quite big.
I've also talked and asked about marking the package as unstable until this issue is fixed, or some other solution is proposed.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
--с уважением, Дозоров Владимир Александрович

rhegner commented:
> and should return to 6.0.1 after all.

6.0.0 was actually the last version that did proper error handling, see my test results above:
https://github.com/minio/minio-dotnet/pull/1141#issuecomment-2255440591

ebozduman commented:
@rhegner 
Could you run your tests if you still have them and let us know your findings.
Appreciate your help on this long belated fix.
This fix brings quite many changes including the functional test fixes. The reason for it is the build tests are needed to run successfully to qualify the missing exception and error handling issues.
There are some scenarios, you'll notice in your tests if you can run, where the exception caught does not explain the root problem clearly. Those can be addressed later. The purpose of this bug was not to miss any exceptions.

xlegalles commented:
We just bumped into the bug https://github.com/minio/minio-dotnet/issues/1118 (supposed to be closed) with a very simple reproducible step: just use wrong credentials and this method will return true. I would be curious to see a test proving that this is an expected behavior? We are currently evaluating this code, and frankly, this is quite concerning, especially since the pull request has been open for a year.

ebozduman commented:
@xlegalles,

There is a fix for the credentials issue, which is currently under testing, will be ready as soon as this PR is merged.

ebozduman commented:
@Willimaend
still in code review phase

ebozduman commented:
@sivanantham-gnanavel 
Could you please code review this PR?
This is an important and long awaited fix.

IgorAtUnisys commented:
We are waiting patiently for this PR! I really hope more ressources are put into this.

ebozduman commented on Minio/MinioClient.cs:
```diff
@@ -119,16 +119,32 @@ internal static void ParseError(ResponseResult response)
-            || HttpStatusCode.MethodNotAllowed == response.StatusCode
-            || HttpStatusCode.NotImplemented == response.StatusCode)
-            ParseWellKnownErrorNoContent(response);
+        var statusCodeStrs = new[]
```
Done!
Will resolve the conversation after committing.

ebozduman commented on Minio/MinioClient.cs:
```diff
@@ -119,16 +119,32 @@ internal static void ParseError(ResponseResult response)
+            ParseWellKnownErrorNoContent(response);
+        }
 #pragma warning disable MA0099 // Use Explicit enum value instead of 0
         if (response.StatusCode == 0)
```
Done.
Will resolve the conversation after committing.

ebozduman commented on Minio/MinioClient.cs:
```diff
@@ -168,22 +185,22 @@ private static void ParseWellKnownErrorNoContent(ResponseResult response)
                 var objectName = resourceSplits[1];
                 errorResponse.Code = "NoSuchKey";
-                error = new ObjectNotFoundException(objectName, "Not found.");
+                error = new ObjectNotFoundException(objectName, "Object NotFound.");
```
@sivanantham-gnanavel 

I'd consider this review/request is out of scope for this PR.

If you'd like, you can file it as a separate issue . 
I think this is a design change. We can easily guess that the goal here was here to give the user ability to customize the exception messages.
So, do you want to take away the users' ability to customize exception messages?

sivanantham-gnanavel replied:
@ebozduman I just want to add a default fallback message so we don’t have to pass the same message repeatedly. I believe you wrote the code that relies on the “NotFound” string if I’m not mistaken.

ebozduman replied:
Default fallback now is "Bucket NotFound" and "Object NotFound"

Willimaendu commented:
Any news?

ebozduman commented on Minio.Functional.Tests/FunctionalTest.cs:
```diff
@@ -371,21 +371,29 @@ internal static async Task BucketExists_Test(IMinioClient minio)
 
             var found = await minio.BucketExistsAsync(beArgs).ConfigureAwait(false);
-            Assert.IsFalse(found);
+            if (found)
```
Yes, Assert functionality also comes with the message field.
But I chose to use `Exceptions` since this is a possible failure that can happen on a customer site. The following explanation was behind my choice:
```
Asserts test for conditions that "cannot happen". 
Exceptions are for conditions that "should not happen but do". 
```

ebozduman commented on Minio/ApiEndpoints/BucketOperations.cs:
```diff
@@ -230,15 +228,15 @@ public async IAsyncEnumerable<Item> ListObjectsEnumAsync(ListObjectsArgs args,
 
 #if NET2_0_OR_GREATER
-    var root = await XDocument.LoadAsync(responseResult.ContentStream, LoadOptions.None, ct).ConfigureAwait(false);
+    var root = await XDocument.LoadAsync(modifiedStream, LoadOptions.None, ct).ConfigureAwait(false);
```
Good catch.
You are right.
Fixed it.

ebozduman commented on Minio/ApiEndpoints/BucketOperations.cs:
```diff
@@ -125,19 +130,13 @@ public async Task RemoveBucketAsync(RemoveBucketArgs args, CancellationToken can
-        if (string.Equals(args.Location, "us-east-1", StringComparison.OrdinalIgnoreCase) &&
-            !string.IsNullOrEmpty(Config.Region))
-            args.Location = Config.Region;
+        if (!string.IsNullOrEmpty(args.Location)) Config.Region = args.Location;
```
`args.Location = Config.Region;` is correct.

ebozduman commented on Minio/ApiEndpoints/BucketOperations.cs:
```diff
@@ -71,26 +77,25 @@ await this.ExecuteTaskAsync(ResponseErrorHandlers, requestMessageBuilder,
+
+            if (response?.Exception is null && response.StatusCode == HttpStatusCode.OK)
+                return true;
+            return false;
```
Done.

ebozduman commented on Minio/ApiEndpoints/BucketOperations.cs:
```diff
@@ -51,15 +52,20 @@ public async Task<ListAllMyBucketsResult> ListBucketsAsync(
         {
-            using var stream = response.ContentBytes.AsStream();
-            bucketList = Utils.DeserializeXml<ListAllMyBucketsResult>(stream);
+            var stream = response.ContentBytes.AsStream();
```
Incomplete. I was adding `await` to it.
Fixed it.

ebozduman commented:
@sivanantham-gnanavel 
Could you please respond or mark the changes done for your review comments as "Resolved"?

sivanantham-gnanavel commented:
> @sivanantham-gnanavel Could you please respond or mark the changes done for your review comments as "Resolved"?

@ebozduman Two comments are marked done but I don't see the fixes. Could you help me?

sivanantham-gnanavel reviewed (approved):
Reviewed the code flow and marking it as LGTM. Leaving the approach review to others.

rhegner commented:
Hi everyone, I just ran my testbench with v6.0.5, hoping the exception handling is finally fixed.

Unfortunately it is still broken compared to 6.0.0:

```
// Minio Version                                6.0.0   6.0.1   6.0.2   6.0.3   6.0.4   6.0.5
// ------------------------------------------------------------------------------------------
await TestExistsWithExistingObject();       //  ok      ok      ok      ok      ok      ok
await TestExistsWithNonexistingObject();    //  ok      NOK     ok      ok      ok      ok
await TestExistsFromInvalidEndpoint();      //  ok      ok      NOK     NOK     NOK     NOK
await TestDownloadWithExistingObject();     //  ok      ok      ok      ok      ok      ok
await TestDownloadWithNonexistingObject();  //  ok      ok      ok      ok      ok      ok
await TestDownloadFromInvalidEndpoint();    //  ok      ok      NOK     NOK     NOK     NOK
await TestWithInvalidCredentials();         //  ok      ok      NOK     NOK     NOK     NOK
```

In the two NOK cases with an invalid endpoint, the library now throws a `Minio.Exceptions.InternalClientException`, whereas it used to throw a `Minio.Exceptions.ConnectionException` with v6.0.0, which is a breaking change.

In the last case with invalid credentials, the library now throws a `System.NullReferenceException`, whereas it used to throw a `Minio.Exceptions.AccessDeniedException` with v6.0.0, which is a breaking change.

ebozduman commented:
Thank you @rhegner 

You are right. The behavior changed, so the version number was supposed to be 7.0 and the behavior change should have been included in it's Release Notes.

I think the new/current exception thrown for an invalid endpoint is better than the previous "ConnectionException".
```
Unhandled exception. Minio.Exceptions.InternalClientException: MinIO API responded with message=Unsuccessful response from server without XML:Name or service not known (play.mina.io:443). Status code=OK, response=Name or service not known (play.mina.io:443), content=
```

When it comes to access and secret keys though, the returned exception for incorrect values is way too generic.
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
```

The reason why a new release is generated without the fix for invalid access/secret key problem was to be able to deliver the main and much more important fix for returning fake `"exists"` status for buckets and objects.

So, this is a known problem that we need to fix and the root cause needs to be explained in the exception message.
